README.md

```diff
@@ -1,161 +1,720 @@
+
+There MUST be at least one genesis node in an inhabited graph (even if it is a single-node graph). There MAY be an unbounded number of concurrent genesis nodes.
+
+#### 4.2.2.2 Head Nodes
+
+The "latest" facts in the above graph are `bafy...berry`, and `bafy...coffee` (highlighted in green). These are called "head nodes", as they have no known descendants at read-time. In formal terms, these are "causal sources". These nodes are merely subjectively the most recent fact: it is possible that some other facts were written elsewhere, but simply have not arrived at the reader yet. Since more facts MAY be appended to the history at any time.
+
+There MUST be at least one head node in an inhabited graph (even if it is a single-node graph). There MAY be an unbounded number of concurrent head nodes.
+
+#### 4.2.2.3 Concurrent Nodes
+
+When comparing any two nodes, if one does not appear in the causal history of the other, then they are said to be "concurrent" or "sibling" nodes. As any two causally disjoint graphs have (by definition) no dependencies on each other, most facts are concurrent.
+
+#### 4.2.2.4 Parent Nodes
+
+Nodes listed in the [`Cause`] field of a fact are said to be that fact's "parent nodes". A fact MUST have zero or more parents.
+
+``` mermaid
+flowchart BT
+    parentA
+    parentB
+    
+    fact --> parentA
+    fact --> parentB
+```
+
+#### 4.2.2.5 Child Nodes
+
+Nodes that list a fact is their [`Cause`] field of a fact are said to be that fact's "child nodes". A fact MUST have zero or more children.
+
+``` mermaid
+flowchart BT
+    childA --> fact
+    childB --> fact
+```
 
-## 2.10 Storage
+#### 4.2.2.6 Ancestors & Provenance
 
-TODO: Introduce + link Brooke's upcoming work on persistence + encryption
\ No newline at end of file
+A complete causal history is built up by recursively following parent edges, from the node being investigated back to its geneses. As long as there is an unbroken path from one node to another, it is said to be the "descendant" of its "ancestor". For example, in the above graph, `bafy...avocado` is an ancestor of `bafy...baklava` along the blue path.
+
+Only direct parents SHOULD be listed in a [`Cause`] field, as the complete history is intact [transatively][transative]. For example, in the graph above, `bafy...ambrosia` has no direct link to `bafy...agave` and `bafy...bun` has no direct link to `bafy...bean` because indirect, transative histories exists (shown in pink and orange respectively). The fact that this path crosses writers or stores is immaterial.
+
+One exception to avoiding writing redundant links in a causal history when some of those ancestors are expected to have different visibility to peers, such as when some facts are encrypted. 
```

Comment:
Do you have a concrete example of what you mean here? Everything I can think of is better modeled as an unencrypted envelop for an encrypted payload, which avoids the need to write redundant links

Reply:
Sure! Below is a bit of a contrived example to keep the storytelling simple, but the shape is right:

For Entity 123, all updates to country of residence are encrypted, but all surname changes are public data. Alice can read all facts with a "surname" facts, but Bob cannot.

Alice writes a `country` update on top of another updating the entity's name. Bob can read the name update, but it's disconnected from the rest of the DAG, and thus he's unable to tell if it's before or after the previous update he's seen.

<img width="998" alt="Screenshot 2023-06-20 at 18 08 19" src="https://github.com/RhizomeDB/spec/assets/1052016/c09ad525-7d51-4c7a-a265-d8cd34602b24">
<img width="569" alt="Screenshot 2023-06-20 at 18 08 21" src="https://github.com/RhizomeDB/spec/assets/1052016/8c6968ff-b39c-4aa9-b790-1fb803736375">

If you know that only some of the data is visible, you can include a redundant link to the names and heal this for those that can't read your encrypted data.

<img width="998" alt="Screenshot 2023-06-20 at 18 10 38" src="https://github.com/RhizomeDB/spec/assets/1052016/4ac6a186-df60-4ba8-bf87-18b4e1689413">

Reply:
By an unencrypted envelope, do you mean structuring the data like this:

```
[encrypt(key1, entityID), encrypt(key1, attribute), encrypt(key1, value), [cid1, cid2, cid3]]
```

If so, that structure doesn't work under data partition (which we've historically treated as being equal to encryption). For example, if the reader has access to only one store, but the updates are spread across multiple, you doens't have anything clearetext to anchor to.

Reply:
> By an unencrypted envelope, do you mean structuring the data like this:

Not quite! I mean instead having two facts for each write: one unencrypted fact whose value is the CID of the encrypted fact, with a causal link pointing to the fact its anchored off of, and one encrypted fact that contains the actual information.

Something like:

```
C0 = ...
C1 = [entityID, "updatedBy", C2, [C0]] (unencrypted)
C2 = [entityID, "country", "UK", []] (encrypted)
```

Reply:
Ah, okay, I think this is one of the thigs we talked past each other about a few weeks ago, then. I thought that you _didn't_ want blank nodes for this kind of things. That makes sense to me for a lot of use cases!

Just to clarify: is that an additional cleatext fact for every cipherfact? Isn't that a lot of extra data, or is it a special kind of more compact node?

```mermaid
flowchart RL
    subgraph clear
        e --> d --> c --> anchor3 --> anchor2 --> b --> anchor1 --> a
    end

    subgraph encrypted
        enc3 --> anchor3
        enc2 --> anchor2
        enc1 --> anchor1
    end
```

How would you model that in the case where you're collaborating with external stores / writers? I'm not sure that you can drop an anchor for every write since these are all single writer stores (so presumably you can't drop anchors on both sides)?

```mermaid
flowchart RL

    subgraph alice
        direction RL
        subgraph clear
            anchor1
            anchor2
            anchor3
            a
            b
            c
            d
            e
            x
            z
        end

        subgraph encrypted
            enc1
            enc2
            enc3
        end
    end

    subgraph bob
        w
        y
    end

    e --> d --> c --> anchor3 --> anchor2 --> b --> anchor1 --> a
    x --> w

    w --> e
    y --> x

    z --> y

    enc1 --> anchor1
    enc2 --> anchor2
    enc3 --> anchor3
```

Reply:
I did a quick sketch of the anchor version on multiple stores, and I must be misunderstanding something. My current version has multiple links kind of like in my original proposal:

```mermaid
flowchart RL

    subgraph alice
        direction RL
        subgraph clear
            anchor1
            anchor2
            anchor3
            a
            b
            c
            d
            e
            x
            z
            anchor4
            anchor6
        end

        subgraph encrypted
            enc1
            enc2
            enc3
        end
    end

    subgraph bob
        w
        y

        anchor5
    end

    e --> d --> c --> anchor3 --> anchor2 --> b --> anchor1 --> a
    x --> w

    w --> e
    %% y --> x

    %% z --> y

    enc1 --> anchor1
    enc2 --> anchor2
    enc3 --> anchor3

    y --> anchor5 --> w
    anchor5 --> x

    x --> anchor4
    anchor4 --> e
    z --> anchor6 --> x
    anchor6 --> y
```

Reply:
To keep it consistent, my proposal in the spec looks like this:

```mermaid
flowchart RL

    subgraph alice
        direction RL
        subgraph clear
            %% anchor1
            %% anchor2
            %% anchor3
            a
            b
            c
            d
            e
            x
            z
            %% anchor4
            %% anchor6
        end

        subgraph encrypted
            enc1
            enc2
            enc3
        end
    end

    subgraph bob
        w
        y

        %% anchor5
    end

    e --> d --> c 
    %% --> anchor3 --> anchor2 --> b --> anchor1 --> a
    x --> w

    w --> e
    %% y --> x

    %% z --> y

    enc1 --> a
    b --> a
    b --> enc1

    enc2 --> b
    enc3 --> enc2
    c --> b
    c --> enc3

    x --> e
    y --> w
    z --> y
    z --> x

    %% enc1 --> anchor1
    %% enc2 --> anchor2
    %% enc3 --> anchor3

    %% y --> anchor5 --> w
    %% anchor5 --> x

    %% x --> anchor4
    %% anchor4 --> e
    %% z --> anchor6 --> x
    %% anchor6 --> y
```